crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -528,6 +532,9 @@ where
     fn visit_stmt(&mut self, stmt: &'ast ast::Stmt) {
         match stmt {
             ast::Stmt::FunctionDef(function_def) => {
+                let definition_inside_loop = self.inside_loop;
+                self.inside_loop = false;
```

Comment:
would it make more sense to reset this in `push_scope()` and `pop_scope()`, rather than having very similar code in the branches for `Stmt::FunctionDef` and `Stmt::ClassDef`?

Reply:
Doing this would fix @MichaReiser's comment in https://github.com/astral-sh/ruff/pull/14522#discussion_r1853665736

Reply:
Hmm, I see your point.

I think the design of the builder here would be nicer to deal with if the pushing and popping of the scope happened in the same method -- something more similar to the `with_type_params` method and the `with_generators_scope` methods that we already have. We'd have to keep track of fewer stacks on the builder that deal with issues of nested scopes, and it would make it easier to maintain the invariant that a `push_scope` call should always be balanced with a `pop_scope` call.

For now I think I'm okay with what you have. The implicit invariant we're relying on here is that class definitions and functions defined via `def` are the only kinds of nested scopes that can contain statements (and therefore `break` statements). This will break if Python adds another kind of scope in a future version that allows statements in its body, but this does seem unlikely.

Reply:
I'm fine with either solution. Using a stack seems complicated for just the one flag.